### PR TITLE
Fix for issue #1261 Build/Research Costs and Times incorrect

### PR DESCRIPTION
**Pull request by LGM-Doyle**
I think that this fixes issue #1261 by reverting the problematic commit tracked down by @dbenage-cx .

I also expanded the error messages in the log, so that missing but expected sources show up as errors and not just weird default values.

### CODE REVIEW

File: universe/Building.cpp [hunk 1]
```diff
@@ -239,13 +239,20 @@ float BuildingType::ProductionCost(int empire_id, int location_id) const {
         if (m_production_cost && m_production_cost->ConstantExpr())
             return m_production_cost->Eval();
 
+        const auto arbitrary_large_number = 999999.9f;
+
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
-        if (!location)
-            return 999999.9f;    // arbitrary large number
+        if (!location) {
+            ErrorLogger() << "Location is missing using production cost of " << arbitrary_large_number;
```
Comment (geoffthemedio) [inline universe/Building.cpp]: not necessarily an error... pedia can show stuff even when there's no game ongoing.

Comment (LGM-Doyle) [reply universe/Building.cpp]: Is there a way to determine if the pedia is running off-line?

If so, I could add an earlier check `if (offline ) return arbitrary_large_number;` that does not generate an error message.

**Comment by geoffthemedio** [comment]
even if it's in a game, the pedia value probably won't have a valid source. just don't treat it as an error.

**Comment by Cjkjvfnby** [comment]
`Fix 1261 build times incorrect` did you cont all 1261 build times?   

PS Please add `#` to issue reference to avoid confusions.

**Comment by LGM-Doyle** [comment]
I've removed the error messages on the Location clauses in Buildings.cpp.

**Comment by o01eg** [comment]
What if use boost::optional instead special values?

**Comment by geoffthemedio** [comment]
Then there would need to be a check later for whether the optional value wasn't provided, and in that case use arbitrary values or some other way to display something in the UI.

File: universe/Building.cpp [hunk 2]
```diff
@@ -273,10 +271,8 @@ int BuildingType::ProductionTime(int empire_id, int location_id) const {
         const auto arbitrary_large_number = 9999;
 
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
-        if (!location) {
+        if (!location)
             ErrorLogger() << "Location is missing using production time of " << arbitrary_large_number;
-            return arbitrary_large_number;
```
Comment (dbenage-cx) [inline universe/Building.cpp]: Did you mean to remove the previous line instead of this one?

Comment (LGM-Doyle) [reply universe/Building.cpp]: Fixed.

File: universe/Building.cpp [hunk 3]
```diff
@@ -239,13 +239,18 @@ float BuildingType::ProductionCost(int empire_id, int location_id) const {
...
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
         if (!location)
-            return 999999.9f;    // arbitrary large number
+            return arbitrary_large_number;
 
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
-        if (!source || m_production_cost->SourceInvariant())
-            return 999999.9f;
+        if (!source && !m_production_cost->SourceInvariant()) {
+            ErrorLogger() << "Source is expected but missing using production cost of "
```
Comment (geoffthemedio) [inline universe/Building.cpp]: same here... not always an error / unexpected.

the message is also poorly worded / ambiguous

File: universe/Building.cpp [hunk 4]
```diff
@@ -263,13 +268,18 @@ int BuildingType::ProductionTime(int empire_id, int location_id) const {
...
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
         if (!location)
-            return 9999;    // arbitrary large number
+            return arbitrary_large_number;
 
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
-        if (!source || m_production_time->SourceInvariant())
-            return 9999;
+        if (!source && !m_production_time->SourceInvariant()) {
+            ErrorLogger() << "Source is expected but missing using production time of "
```
Comment (geoffthemedio) [inline universe/Building.cpp]: ...

File: universe/ShipDesign.cpp [hunk 5]
```diff
@@ -330,13 +330,20 @@ float PartType::ProductionCost(int empire_id, int location_id) const {
         if (m_production_cost->ConstantExpr())
             return static_cast<float>(m_production_cost->Eval());
 
+        const auto arbitrary_large_number = 999999.9f;
+
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
-        if (!location)
-            return 999999.9f;    // arbitrary large number
+        if (!location) {
+            ErrorLogger() << "Location is missing using production cost of " << arbitrary_large_number;
```
Comment (geoffthemedio) [inline universe/ShipDesign.cpp]: ...

File: universe/ShipDesign.cpp [hunk 6]
```diff
@@ -330,13 +330,20 @@ float PartType::ProductionCost(int empire_id, int location_id) const {
...
+            return arbitrary_large_number;
+        }
 
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
-        if (!source || m_production_cost->SourceInvariant())
-            return 999999.9f;
+        if (!source && !m_production_cost->SourceInvariant()) {
+            ErrorLogger() << "Source is expected but missing using production cost of "
```
Comment (geoffthemedio) [inline universe/ShipDesign.cpp]: ...

File: universe/ShipDesign.cpp [hunk 7]
```diff
@@ -345,19 +352,26 @@ float PartType::ProductionCost(int empire_id, int location_id) const {
...
         return 1;
     } else {
         if (m_production_time->ConstantExpr())
             return m_production_time->Eval();
 
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
-        if (!location)
-            return 9999;    // arbitrary large number
+        if (!location) {
+            ErrorLogger() << "Location is missing using production time of " << arbitrary_large_number;
```
Comment (geoffthemedio) [inline universe/ShipDesign.cpp]: ...

File: universe/ShipDesign.cpp [hunk 8]
```diff
@@ -345,19 +352,26 @@ float PartType::ProductionCost(int empire_id, int location_id) const {
...
+            return arbitrary_large_number;
+        }
 
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
-        if (!source || m_production_time->SourceInvariant())
-            return 9999;
+        if (!source && !m_production_time->SourceInvariant()) {
+            ErrorLogger() << "Source is expected but missing using production time of "
```
Comment (geoffthemedio) [inline universe/ShipDesign.cpp]: ...

**Review by LGM-Doyle — COMMENTED** [review]
I've removed 2 error messages in Buildings.cpp and 4 in ShipDesign.cpp.  

I've reworded the missing source error message in Tech.cpp.

File: universe/ShipDesign.cpp (hunk 5 above)
Comment (geoffthemedio) [reply universe/ShipDesign.cpp]: really? I still see it in the diff

File: universe/Tech.cpp [hunk 9]
```diff
@@ -160,8 +160,8 @@ float Tech::ResearchCost(int empire_id) const {
     } else {
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
         if (!source && !m_research_cost->SourceInvariant()) {
-            ErrorLogger() << "Source is expected but missing using research cost of "
-                          << arbitrary_large_number;
+            ErrorLogger() << "Using default research cost of " << arbitrary_large_number
```
Comment (geoffthemedio) [inline universe/Tech.cpp]: I'm also skeptical about logging an error message here for the same reasons...

**Comment by LGM-Doyle** [comment]
I've removed all of the error messages.

I assumed that the ones you had not itemized were acceptable.

**Pull request merged**